source/framework/vx_image.c

```diff
@@ -376,6 +382,300 @@ static vx_status ownAllocImageBuffer(vx_reference ref)
+            {
+                status = vxCopyImagePatch(output, &rect, i, &addr, ptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST);
+            }
+            vxUnmapImagePatch(input, map_id);
+        }
+    }
+    return status;
+}
+
+/*! \brief Adjust the memory pointer of the tensor or image by the given offset
+ * This involves non-recursively visiting every sub-image and sub-tensor
+ * and adjusting the memory pointer for that, as well.
+ */
+static vx_status adjustMemoryPointer(vx_reference ref, uint64_t offset[TIVX_IMAGE_MAX_PLANES])
+{
+    vx_status status = VX_SUCCESS;
+    vx_reference stack[TIVX_IMAGE_MAX_SUBIMAGES];
+    vx_image *subimages = NULL;
+    vx_uint32 stack_pointer = 0;
+    vx_uint32 i;
+
+    for (i = 0; i < TIVX_IMAGE_MAX_SUBIMAGES; ++i)
+    {
+        stack[i] = NULL;
+    }
+    stack[stack_pointer++] = ref;
+    while (0 < stack_pointer)
+    {
+        ref = stack[--stack_pointer];
+        if (VX_TYPE_IMAGE == ref->type)
+        {
+            subimages = ((vx_image)ref)->subimages;
+            tivx_obj_desc_image_t *obj_desc = (tivx_obj_desc_image_t *)ref->obj_desc;
+            if (obj_desc->planes > 0)
+            {
+                for (i = 0; i < obj_desc->planes; ++i)
+                {
+                    obj_desc->mem_ptr[i].host_ptr = obj_desc->mem_ptr[i].host_ptr + offset[i];
+                    obj_desc->mem_ptr[i].shared_ptr = tivxMemHost2SharedPtr(obj_desc->mem_ptr[i].host_ptr, TIVX_MEM_EXTERNAL);
+                }
+            }
+            else
+            {
+                obj_desc->mem_ptr[0].host_ptr = obj_desc->mem_ptr[0].host_ptr + offset[((vx_image)ref)->channel_plane];
+                obj_desc->mem_ptr[0].shared_ptr = tivxMemHost2SharedPtr(obj_desc->mem_ptr[0].host_ptr, TIVX_MEM_EXTERNAL);
+            }
+        }
+        else
+        {
+            VX_PRINT(VX_ZONE_ERROR, "Incorrect type passed to adjustMemoryPointer()\n");
+            status = VX_ERROR_INVALID_TYPE;
+            break;
+        }
+        for (i = 0; i < TIVX_IMAGE_MAX_SUBIMAGES; ++i)
+        {
+            if (NULL != subimages[i])
+            {
+                if (TIVX_IMAGE_MAX_SUBIMAGES < stack_pointer)
+                {
+                    VX_PRINT(VX_ZONE_ERROR, "Too many sub-images, may need to increase the value of TIVX_SUBOBJECT_STACK_SIZE in include/TI/tivx_config.h\n");
+                    status = VX_ERROR_NO_RESOURCES;
+                    break;
+                }
+                else
+                {
+                    stack[stack_pointer++] = &subimages[i]->base;
+                }
+            }
+        }
+    }
+    return status;
+}
+
+static vx_status swapImage(const vx_image input, const vx_image output)
+{
+    /* Swap image handles. Need to recalculate for all the sub-images
+        and sub-tensors.
+        NOTE that this will need updating if uniform images are
+        implemented in a more efficient manner!
+    */
+    vx_status status = ownReferenceLock(&output->base);
```

Comment:
This should lock the input and output, correct?

Reply:
done